orphan removal unsupported

Hi, we are facing an issue when using `orphanRemoval`, cascade goes into `onDelete` instead of `reactiveOnDelete`, and it throws an `UnsupportedOperationException`

## Example

Product
```
@Entity
@Table(name = "PRODUCT")
public class Product {

    @ManyToOne(targetEntity = Shop.class, cascade = MERGE)
    @JoinColumn(name = "SHOP_ID", referencedColumnName = "ID")
    private Shop shop;
   
    ...
}
```

Shop
```
@Entity
@Table(name = "SHOP")
public class Shop {

    @OneToMany(mappedBy = "shop", cascade = ALL, fetch = EAGER, orphanRemoval = true)
    private List<Product> productList = new ArrayList<>();

    ...
}
```

Main application
```
EntityManagerFactory emf = Persistence.createEntityManagerFactory("mysql");
Stage.SessionFactory sessionFactory = emf.unwrap(Stage.SessionFactory.class);

sessionFactory.withTransaction((session, transaction) -> {

        Shop s1 = Shop.builder().name("s1").build();
        s1.addProduct(Product.builder().name("ap1").build());
        s1.addProduct(Product.builder().name("ap2").build());
        s1.addProduct(Product.builder().name("ap3").build());
        s1.addProduct(Product.builder().name("ap4").build());

        // create
        return session.merge(s1)
                    .thenCombine(session.flush(), (s, v) -> s);
})
.thenCompose(t1 -> {
        return sessionFactory.withTransaction((session, transaction) -> {
            
            // get
            return session.find(Shop.class, t1.getId())
                    .thenCompose(s2 -> {

                        // update
                        s2.setName("s2");
                        s2.getProductList().clear();
                        s2.addProduct(Product.builder().name("bp5").build());
                        s2.addProduct(Product.builder().name("bp6").build());
                        s2.addProduct(Product.builder().name("bp7").build());

                        // update
                        return session.merge(s2);
                    })
                    .thenCombine(session.flush(), (s, v) -> s);
        });
});
```

## Stacktrace

```
java.util.concurrent.CompletionException: javax.persistence.PersistenceException: org.hibernate.HibernateException: javax.persistence.PersistenceException: org.hibernate.HibernateException: java.lang.UnsupportedOperationException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:824)
	at java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:834)
	at java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2155)
	at java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:110)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.doFlush(ReactiveSessionImpl.java:800)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveAutoflush(ReactiveSessionImpl.java:787)
	at org.hibernate.reactive.stage.impl.StageSessionImpl$Transaction.flush(StageSessionImpl.java:409)
	at org.hibernate.reactive.stage.impl.StageSessionImpl$Transaction.lambda$execute$2(StageSessionImpl.java:391)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	...
Caused by: java.lang.UnsupportedOperationException
	at org.hibernate.reactive.event.impl.DefaultReactiveDeleteEventListener.onDelete(DefaultReactiveDeleteEventListener.java:86)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:113)
	at org.hibernate.internal.SessionImpl.fireDelete(SessionImpl.java:924)
	... 93 more
```

## Sample application

https://github.com/pqab/orphanRemoval-example

## Environment Information

- Version: 1.0.0.Alpha11

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/ReactiveActionQueue.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/ReactiveActionQueue.java
@@ -47,18 +47,18 @@ public class ReactiveActionQueue {
 	static {
 		EXECUTABLE_LISTS_MAP = new LinkedHashMap<>( 8 );
 
-//		EXECUTABLE_LISTS_MAP.put(
-//				OrphanRemovalAction.class,
-//				new ListProvider<OrphanRemovalAction>() {
-//					ExecutableList<OrphanRemovalAction> get(ReactiveActionQueue instance) {
-//						return instance.orphanRemovals;
-//					}
-//					ExecutableList<OrphanRemovalAction> init(ReactiveActionQueue instance) {
-//						// OrphanRemovalAction executables never require sorting.
-//						return instance.orphanRemovals = new ExecutableList<>( false );
-//					}
-//				}
-//		);
+		EXECUTABLE_LISTS_MAP.put(
+				ReactiveOrphanRemovalAction.class,
+				new ListProvider<ReactiveOrphanRemovalAction>() {
+					ExecutableList<ReactiveOrphanRemovalAction> get(ReactiveActionQueue instance) {
+						return instance.orphanRemovals;
+					}
+					ExecutableList<ReactiveOrphanRemovalAction> init(ReactiveActionQueue instance) {
+						// OrphanRemovalAction executables never require sorting.
+						return instance.orphanRemovals = new ExecutableList<>( false );
+					}
+				}
+		);
 		EXECUTABLE_LISTS_MAP.put(
 				ReactiveEntityInsertAction.class,
 				new ListProvider<ReactiveEntityInsertAction>() {
@@ -180,7 +180,7 @@ public class ReactiveActionQueue {
 	private ExecutableList<ReactiveCollectionRemoveAction> collectionRemovals;
 	// TODO: The removeOrphan concept is a temporary "hack" for HHH-6484.  This should be removed once action/task
 	// ordering is improved.
-	private ExecutableList<OrphanRemovalAction> orphanRemovals;
+	private ExecutableList<ReactiveOrphanRemovalAction> orphanRemovals;
 	private transient boolean isTransactionCoordinatorShared;
 	private AfterTransactionCompletionProcessQueue afterTransactionProcesses;
 	private BeforeTransactionCompletionProcessQueue beforeTransactionProcesses;
@@ -397,9 +397,9 @@ public class ReactiveActionQueue {
 	 *
 	 * @param action The action representing the orphan removal
 	 */
-	public void addAction(OrphanRemovalAction action) {
-//		addAction( OrphanRemovalAction.class, action );
-		throw new UnsupportedOperationException();
+	public void addAction(ReactiveOrphanRemovalAction action) {
+		addAction( ReactiveOrphanRemovalAction.class, action );
+//		throw new UnsupportedOperationException();
 	}
 
 	/**

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/Cascade.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/Cascade.java
@@ -11,16 +11,27 @@ import org.hibernate.HibernateException;
 import org.hibernate.bytecode.enhance.spi.interceptor.LazyAttributeLoadingInterceptor;
 import org.hibernate.collection.spi.PersistentCollection;
 import org.hibernate.engine.internal.CascadePoint;
-import org.hibernate.engine.spi.*;
+import org.hibernate.engine.spi.CascadeStyle;
+import org.hibernate.engine.spi.CollectionEntry;
+import org.hibernate.engine.spi.EntityEntry;
+import org.hibernate.engine.spi.PersistenceContext;
+import org.hibernate.engine.spi.Status;
 import org.hibernate.event.spi.EventSource;
 import org.hibernate.internal.CoreLogging;
 import org.hibernate.internal.CoreMessageLogger;
+import org.hibernate.internal.util.collections.IdentitySet;
 import org.hibernate.persister.collection.CollectionPersister;
 import org.hibernate.persister.entity.EntityPersister;
 import org.hibernate.pretty.MessageHelper;
 import org.hibernate.proxy.HibernateProxy;
 import org.hibernate.reactive.session.ReactiveSession;
-import org.hibernate.type.*;
+import org.hibernate.reactive.util.impl.CompletionStages;
+import org.hibernate.type.AssociationType;
+import org.hibernate.type.CollectionType;
+import org.hibernate.type.CompositeType;
+import org.hibernate.type.EntityType;
+import org.hibernate.type.ForeignKeyDirection;
+import org.hibernate.type.Type;
 
 import java.io.Serializable;
 import java.util.Collection;
@@ -548,11 +559,13 @@ public final class Cascade<C> {
 			orphans = pc.getQueuedOrphans( entityName );
 		}
 
-		for ( Object orphan : orphans ) {
+		ReactiveSession session = (ReactiveSession) eventSource;
+		stage = stage.thenCompose( v -> CompletionStages.loop( orphans, orphan -> {
 			if ( orphan != null ) {
 				LOG.tracev( "Deleting orphaned entity instance: {0}", entityName );
-				eventSource.delete( entityName, orphan, false, new HashSet<>() );
+				return session.reactiveRemove( orphan, false, new IdentitySet() );
 			}
-		}
+			return voidFuture();
+		} ) );
 	}
 }

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveOrphanRemovalAction.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/ReactiveOrphanRemovalAction.java
@@ -1,0 +1,19 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.engine.impl;
+
+import org.hibernate.engine.spi.SessionImplementor;
+import org.hibernate.persister.entity.EntityPersister;
+
+import java.io.Serializable;
+
+public class ReactiveOrphanRemovalAction extends  ReactiveEntityDeleteAction {
+    public ReactiveOrphanRemovalAction(Serializable id, Object[] state, Object version, Object instance,
+                                       EntityPersister persister, boolean isCascadeDeleteEnabled,
+                                       SessionImplementor session) {
+        super(id, state, version, instance, persister, isCascadeDeleteEnabled, session);
+    }
+}

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveDeleteEventListener.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveDeleteEventListener.java
@@ -8,7 +8,6 @@ package org.hibernate.reactive.event.impl;
 import org.hibernate.HibernateException;
 import org.hibernate.LockMode;
 import org.hibernate.TransientObjectException;
-import org.hibernate.action.internal.OrphanRemovalAction;
 import org.hibernate.engine.internal.CascadePoint;
 import org.hibernate.reactive.engine.impl.ForeignKeys;
 import org.hibernate.engine.internal.Nullability;
@@ -29,6 +28,7 @@ import org.hibernate.jpa.event.spi.CallbackRegistryConsumer;
 import org.hibernate.persister.entity.EntityPersister;
 import org.hibernate.pretty.MessageHelper;
 import org.hibernate.reactive.engine.ReactiveActionQueue;
+import org.hibernate.reactive.engine.impl.ReactiveOrphanRemovalAction;
 import org.hibernate.reactive.session.ReactiveSession;
 import org.hibernate.reactive.engine.impl.Cascade;
 import org.hibernate.reactive.engine.impl.CascadingActions;
@@ -360,7 +360,7 @@ public class DefaultReactiveDeleteEventListener
 						// TODO: The removeOrphan concept is a temporary "hack" for HHH-6484.  This should be removed once action/task
 						// ordering is improved.
 						actionQueue.addAction(
-								new OrphanRemovalAction(
+								new ReactiveOrphanRemovalAction(
 										entityEntry.getId(),
 										deletedState,
 										version,

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ReactiveResultSetProcessor.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ReactiveResultSetProcessor.java
@@ -14,9 +14,12 @@ import org.hibernate.internal.CoreLogging;
 import org.hibernate.internal.CoreMessageLogger;
 import org.hibernate.loader.plan.exec.query.spi.NamedParameterContext;
 import org.hibernate.loader.spi.AfterLoadAction;
+import org.hibernate.persister.entity.EntityPersister;
 import org.hibernate.reactive.session.ReactiveSession;
 import org.hibernate.reactive.util.impl.CompletionStages;
 import org.hibernate.transform.ResultTransformer;
+import org.hibernate.type.EntityType;
+import org.hibernate.type.OneToOneType;
 
 import java.io.Serializable;
 import java.sql.ResultSet;
@@ -60,10 +63,10 @@ public interface ReactiveResultSetProcessor {
 				entity,
 				entityEntry,
 				session,
-				(entityType, value, session1, owner, overridingEager)
+				(entityType, value, source, owner, overridingEager)
 						-> entityType.isEager( overridingEager )
-								? ((ReactiveSession) session1).reactiveGet( entityType.getReturnedClass(), (Serializable) value )
-								: entityType.resolve(value, session1, owner, overridingEager)
+								? resolve( entityType, (Serializable) value, owner, source )
+								: entityType.resolve(value, source, owner, overridingEager)
 		);
 
 		final Object[] hydratedState = entityEntry.getLoadedState();
@@ -83,5 +86,51 @@ public interface ReactiveResultSetProcessor {
 						listeners
 				)
 		);
+	}
+
+	/**
+	 * Replacement for {@link EntityType#resolve(Object, SharedSessionContractImplementor, Object, Boolean)}
+	 */
+	default CompletionStage<Object> resolve(EntityType entityType, Serializable value, Object owner,
+											SharedSessionContractImplementor session) {
+		if ( value != null && !isNull( entityType, owner, session ) ) {
+			if ( entityType.isReferenceToPrimaryKey() ) {
+				return ((ReactiveSession) session).reactiveInternalLoad(
+						entityType.getAssociatedEntityName(),
+						value,
+						true,
+						entityType.isNullable()
+				);
+			}
+			else {
+				//TODO see EntityType.resolve() we need to reactify loadByUniqueKey()
+				throw new UnsupportedOperationException("unique key property name not supported here");
+			}
+		}
+		else {
+			return null;
+		}
+	}
+
+	default boolean isNull(EntityType entityType, Object owner,
+						   SharedSessionContractImplementor session) {
+		if ( entityType instanceof OneToOneType ) {
+			OneToOneType type = (OneToOneType) entityType;
+			String propertyName = type.getPropertyName();
+			if ( propertyName != null ) {
+				EntityPersister ownerPersister =
+						session.getFactory().getMetamodel()
+								.entityPersister( entityType.getAssociatedEntityName() );
+				Serializable id = session.getContextEntityIdentifier(owner);
+				EntityKey entityKey = session.generateEntityKey(id, ownerPersister);
+				return session.getPersistenceContextInternal().isPropertyNull( entityKey, propertyName);
+			}
+			else {
+				return false;
+			}
+		}
+		else {
+			return false;
+		}
 	}
 }

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/ReactiveSession.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/ReactiveSession.java
@@ -108,6 +108,8 @@ public interface ReactiveSession extends ReactiveQueryExecutor {
 			Class<T> entityClass,
 			Object... primaryKey);
 
+	CompletionStage<Object> reactiveInternalLoad(String entityName, Serializable id, boolean eager, boolean nullable);
+
 	CompletionStage<Object> reactiveImmediateLoad(String entityName, Serializable id);
 
 	CompletionStage<Void> reactiveInitializeCollection(PersistentCollection collection, boolean writing);

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveSessionImpl.java
@@ -16,12 +16,14 @@ import org.hibernate.MappingException;
 import org.hibernate.ObjectDeletedException;
 import org.hibernate.ObjectNotFoundException;
 import org.hibernate.TypeMismatchException;
+import org.hibernate.UnresolvableObjectException;
 import org.hibernate.action.internal.BulkOperationCleanupAction;
 import org.hibernate.collection.spi.PersistentCollection;
 import org.hibernate.dialect.Dialect;
 import org.hibernate.engine.internal.StatefulPersistenceContext;
 import org.hibernate.engine.query.spi.HQLQueryPlan;
 import org.hibernate.engine.query.spi.sql.NativeSQLQuerySpecification;
+import org.hibernate.engine.spi.EffectiveEntityGraph;
 import org.hibernate.engine.spi.ExceptionConverter;
 import org.hibernate.engine.spi.NamedQueryDefinition;
 import org.hibernate.engine.spi.NamedSQLQueryDefinition;
@@ -183,6 +185,50 @@ public class ReactiveSessionImpl extends SessionImpl implements ReactiveSession,
 		);
 		return fireLoadNoChecks( event, LoadEventListener.IMMEDIATE_LOAD )
 				.thenApply( v -> event.getResult() );
+	}
+
+	@Override
+	public CompletionStage<Object> reactiveInternalLoad(String entityName, Serializable id,
+													   boolean eager, boolean nullable) {
+		final EffectiveEntityGraph effectiveEntityGraph = getLoadQueryInfluencers().getEffectiveEntityGraph();
+		final GraphSemantic semantic = effectiveEntityGraph.getSemantic();
+		final RootGraphImplementor<?> graph = effectiveEntityGraph.getGraph();
+		boolean clearedEffectiveGraph;
+		if ( semantic == null || graph.appliesTo(entityName) ) {
+			clearedEffectiveGraph = false;
+		}
+		else {
+//				log.debug( "Clearing effective entity graph for subsequent-select" );
+			clearedEffectiveGraph = true;
+			effectiveEntityGraph.clear();
+		}
+
+		final LoadEventListener.LoadType type;
+		if ( nullable ) {
+			type = LoadEventListener.INTERNAL_LOAD_NULLABLE;
+		}
+		else {
+			type = eager
+					? LoadEventListener.INTERNAL_LOAD_EAGER
+					: LoadEventListener.INTERNAL_LOAD_LAZY;
+		}
+
+		threadCheck();
+		LoadEvent event =
+				new LoadEvent( id, entityName, true, this, getReadOnlyFromLoadQueryInfluencers() );
+		return fireLoadNoChecks( event, type )
+				.thenApply( v -> {
+					Object result = event.getResult();
+					if ( !nullable ) {
+						UnresolvableObjectException.throwIfNull( result, id, entityName );
+					}
+					return result;
+				} )
+				.whenComplete( (v, x) -> {
+					if ( clearedEffectiveGraph ) {
+						effectiveEntityGraph.applyGraph( graph, semantic );
+					}
+				} );
 	}
 
 	@Override @SuppressWarnings("unchecked")

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrphanRemovalTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrphanRemovalTest.java
@@ -1,0 +1,130 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import io.vertx.ext.unit.TestContext;
+import org.hibernate.cfg.Configuration;
+import org.junit.Test;
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.ManyToOne;
+import javax.persistence.OneToMany;
+import java.util.HashSet;
+import java.util.Set;
+
+import static javax.persistence.CascadeType.PERSIST;
+import static javax.persistence.CascadeType.REMOVE;
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class OrphanRemovalTest extends BaseReactiveTest {
+
+    @Override
+    protected Configuration constructConfiguration() {
+        Configuration configuration = super.constructConfiguration();
+        configuration.addAnnotatedClass( Version.class );
+        configuration.addAnnotatedClass( Product.class );
+        configuration.addAnnotatedClass( Shop.class );
+        return configuration;
+    }
+
+    @Test
+    public void testOrphan(TestContext context) {
+        Shop shop = new Shop("shop");
+        Product product = new Product("ap1", shop);
+        product.addVersion(new Version(product));
+        shop.addProduct(product);
+        shop.addProduct(new Product("ap2", shop));
+        shop.addProduct(new Product("ap3", shop));
+        shop.addProduct(new Product("ap4", shop));
+
+        test(context,
+                getSessionFactory()
+                        .withTransaction((session, transaction) -> session.persist(shop))
+                        .thenCompose(v -> getSessionFactory().withTransaction((session, transaction) -> session.find(Shop.class, shop.id)
+                                .thenCompose( result -> session.fetch(result.products).thenApply(products -> result) )
+                                .thenAccept(result -> {
+                                    // update
+                                    result.products.clear();
+                                    result.addProduct(new Product("bp5", result));
+                                    result.addProduct(new Product("bp6", result));
+                                    result.addProduct(new Product("bp7", result));
+                                })
+                        ))
+                        .thenCompose(v -> getSessionFactory().withTransaction((session, transaction) ->
+                                session.createQuery("select count(*) from Product", Long.class)
+                                        .getSingleResult().thenAccept(result -> context.assertEquals(3l, result))
+                        ))
+                        .thenCompose(v -> getSessionFactory().withTransaction((session, transaction) ->
+                                session.createQuery("select name from Product" )
+                                        .getResultList().thenAccept(list -> assertThat( list ).containsExactlyInAnyOrder( "bp5", "bp6", "bp7"))
+                        ))
+        );
+    }
+
+    @Entity(name = "ProductVersion")
+    public static class Version {
+        @Id @GeneratedValue
+        private long id;
+
+        @ManyToOne
+        private Product product;
+
+        public Version(Product product) {
+            this.product = product;
+        }
+        Version() {}
+    }
+
+    @Entity(name = "Product")
+    public static class Product {
+
+        @Id @GeneratedValue
+        private long id;
+        private String name;
+
+        Product(String name, Shop shop) {
+            this.name = name;
+            this.shop = shop;
+        }
+
+        Product() {}
+
+        @ManyToOne
+        private Shop shop;
+
+        @OneToMany(mappedBy = "product", cascade = {PERSIST, REMOVE})
+        private Set<Version> versions = new HashSet<>();
+
+        public void addVersion(Version version) {
+            versions.add(version);
+        }
+    }
+
+    @Entity(name = "Shop")
+    public static class Shop {
+
+        @Id @GeneratedValue
+        private long id;
+        private String name;
+
+        Shop(String name) {
+            this.name = name;
+        }
+
+        Shop() {}
+
+        @OneToMany(mappedBy = "shop", cascade = {PERSIST, REMOVE}, orphanRemoval = true)
+        private Set<Product> products = new HashSet<>();
+
+        public void addProduct(Product product) {
+            products.add(product);
+        }
+    }
+
+}

